esp/esdllib/README.md

```diff
@@ -0,0 +1,1608 @@
+
+To see the ESDL script context in trace log output, set `Method/@traceLevel=10`. After exiting each entry point the ESP prints out the contents of the context with the label **Entire script context after transforms:**
+
+### Logging
+
+TODO: Add details about configuring logging.
+
+### Service Security
+
+Keeping your DESDL service secure and ensuring proper access to product features is an important part of the process of writing and configuring your service. ESDL integration scripts can assist with enforcing complex authentication and authorization requirments. Please refer to the separate document *README-SECURITY.md* for details.
+
+# Reference
+
+## Common Attributes
+
+| Name      | Usage Notes |
+|-----------|-------------|
+| @name     | Interpreted as a scalar value so quoting is not required. XPath expressions are not supported. String identifier used in trace output. Use is optional, but strongly encouraged. |
+| @target   | XPath expression that evaluates to the target of an operation or destination of an entry points output document node. As an entry point attribute it has these additional characteristics: <br><br> 1. Special macros for the configured back end query `{$query}` and the ESP SOAP request `{$request}` are available to use. <br> 2. The XPath must exist at runtime or an exception will be thrown. A future release will ensure the path exists without throwing an exception. |
+| @value    | A scalar value- XPath expressions are not supported. |
+| @source   | XPath expression giving the location in the script context where data will be read from. As an attribute of an endpoint element it can contain two special macros: `{$query}` for the configured back end query and `{$request}` for the ESP SOAP request.
+
+## Conventions
+
+### Namespaces
+
+The scripting engine requires strict adherence to namespaces correctness. A frequent source of scripting errors is omitting namespaces where they're required or using an incorrect namespace or namespace prefix.
+
+Namespaces referenced in any of your xpaths must be defined in your scripts to match the XML being processed. One exception is that the BackendRequest entry point will automatically detect and define the standard soap namespace ( `xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/"` ) if it is used in your source XML.
+
+When your source XML uses a namespace prefix as part of the tag name, you must also use it in your xpath referencing that node.
+
+If a node defines a namespace without a prefix, you must define that namespace with a prefix in the root of your script. Then you must use the prefix in any xpaths referencing that node. Any children of that node in the source XML must not be referred to with the prefix you defined in your script:
```

Comment:
@timothyklemm Are you saying we support declaring namespaces for all child scripts in the <scripts> element?  If not, yes, we probably should add that at some point, but do we already?  If you are commenting about a desired feature, please make it clear so Terrence doesn't have to sift through them looking for changes to be made to the document.  Or create a wiki somewhere to discuss them.

Reply:
Using default script namespaces to match default content namespaces will work if the two URIs match.  Even for default prefixes, there is an internal URI.  That's what is used for matching, not the prefix.
There are cases where the internal xml has no URI assigned.  I'd have to dig into the library to remember how exactly it happens.  In those cases you still have to be careful how you match it.

Reply:
City is an example where I'd have to walk through the libxml parser and see how the URI is being set.

Reply:
First point questions the documentation. A required namespace can be defined on the node in which it is used, or any ancestor of that node. My objection is to the "root of the script" as the only acceptable location.

Second point questions both documentation and behavior. It's not intuitive why a script can't define a default namespace in this scenario. If it can't, the documentation should ideally attempt to explain why. If it can, the documentation is wrong.

Reply:
I've moved this section under the XPath section, added examples showing the behavior, and removed language that was inaccurate or causing confusion.